### PR TITLE
incr.comp.: Assert that no DepNode is re-opened (see issue #42298).

### PR DESCRIPTION
**Pull request by michaelwoerister**
This PR removes the last occurrence of DepNode re-opening and adds an assertion that prevents our doing so in the future too. The DepGraph should now be guaranteed to be cycle free.

r? @nikomatsakis 

EDIT: Closes https://github.com/rust-lang/rust/issues/42298

### CODE REVIEW

**Review by nikomatsakis — APPROVED** [review]

File: src/librustc/dep_graph/edges.rs [hunk 1]
```diff
@@ -23,6 +23,11 @@ pub struct DepGraphEdges {
     edges: FxHashSet<(DepNodeIndex, DepNodeIndex)>,
     task_stack: Vec<OpenTask>,
     forbidden_edge: Option<EdgeFilter>,
+
+    // A set to help assert that no two tasks use the same DepNode. This is a
+    // temporary measure. Once we load the previous dep-graph as readonly, this
+    // check will fall out of the graph implementation naturally.
+    opened_once: FxHashSet<DepNode>,
```
Comment (nikomatsakis) [inline src/librustc/dep_graph/edges.rs]: maybe we want to make `#[cfg(debug_assertions)]`?

**Comment by nikomatsakis** [comment]
r=me -- but decide if you think it should be `#[cfg]` or not

**Comment by michaelwoerister** [comment]
@bors r=nikomatsakis

Since this should only ever be hit in incremental mode and that is almost exclusively used with debug-assertions-enabled nightlies, I think it's OK to leave it the way it is.

**Comment by kennytm** [comment]
`gnu-incremental` failed to compile `rustc_data_structures`. Legit.

```
[00:19:50] error: internal compiler error: /checkout/src/librustc/dep_graph/edges.rs:107: Re-opened node ConstEval(Fingerprint(12533344459010390675, 9240382652965425921))
[00:19:50] 
[00:19:50] note: the compiler unexpectedly panicked. this is a bug.
[00:19:50] 
[00:19:50] note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports
[00:19:50] 
[00:19:50] note: rustc 1.21.0-dev (189ca70d2 2017-08-03) running on x86_64-unknown-linux-gnu
[00:19:50] 
[00:19:50] thread 'rustc' panicked at 'Box<Any>', /checkout/src/librustc_errors/lib.rs:486:8
[00:19:50] note: Run with `RUST_BACKTRACE=1` for a backtrace.
[00:19:50] 
[00:19:50] error: Could not compile `rustc_data_structures`.
```

**Comment by michaelwoerister** [comment]
@bors r=nikomatsakis

Passes travis now after the fixes. Let's give it another try.

**Comment by kennytm** [comment]
@bors retry #40474 — `x86_64-pc-windows-gnu` failed to download LLVM 

```
Attempting with retry: sh -c rm -f d9e7d2696e41983b6b5a0b4fac604b4e548a84d3.tar.gz &&             curl -sSL -O https://github.com/rust-lang/llvm/archive/d9e7d2696e41983b6b5a0b4fac604b4e548a84d3.tar.gz
gzip: stdin: not in gzip format
```

**Pull request merged**